File: core/src/com/unibo/s3/main_system/characters/BaseCharacter.java
package com.unibo.s3.main_system.characters;

import akka.actor.ActorRef;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.math.Vector2;
import com.unibo.s3.main_system.characters.steer.BaseMovableEntity;
import com.unibo.s3.main_system.characters.steer.CustomLocation;
import org.jgrapht.UndirectedGraph;
import org.jgrapht.alg.NeighborIndex;
import org.jgrapht.graph.DefaultEdge;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Random;

public class BaseCharacter extends BaseMovableEntity implements Character {

    private Color color;
    private final int id;

    private UndirectedGraph<Vector2, DefaultEdge> graph;
    private int nNeighbours = 0;
    private Vector2 currentNode;
    private List<ActorRef> neighbours = new ArrayList<>();
    private List<Vector2> visited = new ArrayList<>();
    private NeighborIndex index;
    private Vector2 currentDestination;

    public BaseCharacter(Vector2 position, int id) {
        super(position);
        this.id = id;
    }

    @Override
    public int getId() {
        return id;
    }

    @Override
    public void setColor(Color color) {
        this.color = color;
    }

    @Override
    public Color getColor() {
        return color;
    }

    public void setGraph(UndirectedGraph<Vector2, DefaultEdge> g){
        this.graph = g;
        currentNode = computeInitialNearestNode();
        index = new NeighborIndex(graph);
        currentDestination = selectRandomDestination();
        System.out.println(log() + "my destination is " + currentDestination);
        setNewDestination(currentDestination);
    }

    public Vector2 getCurrentNode() {
        return currentNode;
    }

    public List<ActorRef> getNeighbours() {
        return neighbours;
    }

    public void setNeighboursList(List<ActorRef> neighbours){
        this.neighbours = neighbours;
    }


    public void addNeighbour(ActorRef neighbour){
        this.neighbours.add(neighbour);
        this.nNeighbours++;
    }

    public void chooseBehaviour(){

        this.currentNode = computeNearest();
        if (currentNode.equals(currentDestination)) {
            System.out.println(log() + "Destination " + currentDestination + " = " + currentNode + " achieved! Choose the next one");
            currentDestination = selectRandomDestination();
        }
        //ora scelgo destinazione casuale tra i vicini, potenzialmente torno indietro
    }

    public void setnNeighbours(int n){
        this.nNeighbours = n;
    }

    public Vector2 getCurrentDestination() {
        return currentDestination;
    }

    public List<Vector2> getInformations(){
        return this.visited;
    }

    public void updateGraph(List<Vector2> colleagueList){
        this.nNeighbours--;
        //update lista
        for(Vector2 v : colleagueList){
            if(!visited.contains(v)){
                visited.add(v);
            }
        }
        if(nNeighbours == 0){
            chooseBehaviour();
        }
    }

    private void setNewDestination(Vector2 destination){
        //setta destinazione
        this.setComplexSteeringBehavior()
                .avoidCollisionsWithWorld()
                .arriveTo(new CustomLocation(destination))
                .buildPriority(true);
    }

    private void chooseNextDestination(){
        //scegli prossimo nodo tra i raggiungibili
    }

    public List<Vector2> computeNeighbours(){
      return index.neighborListOf(currentNode);
    }

    //computo il mio nodo di riferimento
    public Vector2 computeNearest(){
        Vector2 nearest = currentNode;
        float minDistance = getPosition().dst2(new Vector2(nearest.x, nearest.y));
        List<Vector2> list = computeNeighbours();
        for(Vector2 v : list){
            float distance = (v.dst2(getPosition()));
            if(distance < minDistance){
                nearest = v;
                minDistance = distance;
            }
        }
        if(currentNode != nearest){
            discoverNewVertex(nearest);
        }
        currentNode = nearest;
        if(!computeNeighbours().contains(getCurrentDestination())){ //se ho cambiato nodo di riferimento e questo non è collegato alla destinazione la ricalcolo
            currentDestination = selectRandomDestination();
        }
        return nearest;
    }

    private Vector2 computeInitialNearestNode(){
        Vector2 nearest = new Vector2();
        float minDistance = Float.MAX_VALUE;
        for(Vector2 v : graph.vertexSet()){
            float distance = (v.dst2(getPosition()));
            if(distance < minDistance){
                nearest = v;
                minDistance = distance;
            }
        }
        discoverNewVertex(nearest);
        return nearest;
    }

    private Vector2 selectRandomDestination(){
        return (Vector2) index.neighborListOf(currentNode).get(new Random().nextInt(index.neighborListOf(currentNode).size()));
    }

    private void discoverNewVertex(Vector2 nearest){
        visited.add(nearest);
    }

    private String log(){
        return "Agent " + id + ": ";
    }
}

File: core/src/com/unibo/s3/main_system/characters/BaseCharacter.scala
package com.unibo.s3.main_system.characters

import java.util

import akka.actor.ActorRef
import com.badlogic.gdx.ai.steer.behaviors.Arrive
import com.badlogic.gdx.ai.steer.proximities.FieldOfViewProximity
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.math.{MathUtils, Vector2}
import com.unibo.s3.main_system.characters.steer.behaviors._
import com.unibo.s3.main_system.characters.steer.{BaseMovableEntity, CustomLocation}
import org.jgrapht.UndirectedGraph
import org.jgrapht.alg.NeighborIndex
import org.jgrapht.graph.DefaultEdge

import scala.collection.JavaConverters._
import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.util.Random

trait Character {

  /**initial graph setting*/
  def setGraph(g: UndirectedGraph[Vector2, DefaultEdge]): Unit

  /**getting character infos (maybe guard exclusive)*/
  def getInformation: Iterable[Vector2]

  /**graph update (maybe guard exclusive)*/
  def updateGraph(colleagueList: Iterable[Vector2]): Unit

  /**getting sight line length (maybe thief exclusive)*/
  def getSightLineLength : Float

  def getTarget: Option[Target[BaseCharacter]]

}

abstract class BaseCharacter(vector2: Vector2, id : Int) extends BaseMovableEntity(vector2) with Character {
  private[this] var graph: UndirectedGraph[Vector2, DefaultEdge] = _

  private var currentNode : Option[Vector2] = Option[Vector2](new Vector2())
  private var previousNode : Option[Vector2] = Option[Vector2](new Vector2())

  private var visited = mutable.ArrayBuffer[Vector2]()
  private var visitedBuffer = mutable.ArrayBuffer[Vector2]()
  private var index : NeighborIndex[Vector2,DefaultEdge] = _
  private var currentDestination : Option[Vector2] = None
  private val freqChooseNonVisitedFirst = 0.65f

  private val sightLineLength : Float = 15f
  private val fovAngle = 120f
  private val fovRadius = 5f
  private var coneOfView: FieldOfViewProximity[Vector2] = _

  private var arrive: Arrive[Vector2] = new Arrive[Vector2](this)
  arrive.setTarget(new CustomLocation(getPosition))


  coneOfView  = new FieldOfViewProximity[Vector2](
    this, null, fovRadius, MathUtils.degreesToRadians * fovAngle)

  private val randomGenerator = Random

  def getId: Int = id

  def setGraph(g: UndirectedGraph[Vector2, DefaultEdge]): Unit = {
    this.graph = g
    currentNode = computeInitialNearestNode
    index = new NeighborIndex[Vector2, DefaultEdge](graph)
    currentDestination = selectRandomDestination()
    setNewDestination(currentDestination.get)
  }

  def getInformation: Iterable[Vector2] = this.visited

  def updateGraph(colleagueList: Iterable[Vector2]): Unit =
    visitedBuffer ++ colleagueList

  private def mergeCollegueGraph(): Unit = {
    visitedBuffer.foreach(v => if (!visited.contains(v)) visited += v)
    visitedBuffer.clear()
  }

  def getSightLineLength : Float = this.sightLineLength

  def getFieldOfView: FieldOfViewProximity[Vector2] = coneOfView

  private def setNewDestination(destination: Vector2) = {
    arrive.setTarget(new CustomLocation(destination))
    this.setComplexSteeringBehavior()
      .avoidCollisionsWithWorld()
      .add(arrive)
      .buildPriority(true)
    /*this.setComplexSteeringBehavior()
      .avoidCollisionsWithWorld
      .arriveTo(new CustomLocation(destination))
      .buildPriority(true)*/
  }

  private def computeInitialNearestNode = {
    var nearest = None: Option[Vector2]
    var minDistance = Float.MaxValue
    import scala.collection.JavaConversions._
    for (v <- graph.vertexSet) {
      val distance = v.dst2(getPosition)
      if (distance < minDistance) {
        nearest = Some[Vector2](v)
        minDistance = distance
      }
    }
    discoverNewVertex(nearest)
    nearest
  }

  def selectRandomDestination(): Option[Vector2] = {
    if (index != null && currentNode.isDefined) {
      val neighbors = index.neighborListOf(currentNode.get)
      val n = neighbors.size()
      if (n == 1) return Option(neighbors.get(0))
      if (n > 1) return Option(neighbors.get(randomGenerator.nextInt(n)))
    }
    currentNode
  }

  private def selectPriorityDestination : Option[Vector2] = {
    mergeCollegueGraph()

    val _neighbors = index.neighborListOf(currentNode.get).asScala
      .filter(node => !node.equals(previousNode.get))
    val notVisited = _neighbors.filter(n => !visited.contains(n))

    var out : Option[Vector2] = None
    val rand = randomGenerator.nextFloat() <= freqChooseNonVisitedFirst
    val neighbors = if(notVisited.nonEmpty && rand) notVisited else _neighbors
    val n = neighbors.size

    if (neighbors.isEmpty) out = previousNode
    else out = Option(neighbors(randomGenerator.nextInt(n)))
    out
  }

  private def discoverNewVertex(nearest: Option[Vector2]): Unit = {
    this.visited :+= nearest.get
  }

  def getCurrentNode: Option[Vector2] = currentNode

  def chooseBehaviour(): Unit = this.currentNode = computeNearestVertex

  private def computeNeighbours: Option[util.List[Vector2]] = {
    import scala.collection.JavaConversions._
    if (index == null){
      Option(List())
    }else{
      Option(index.neighborListOf(currentNode.get))
    }
  }

  //computo il mio nodo di riferimento
  private def computeNearestVertex: Option[Vector2] = {
    var nearest = currentNode
    var minDistance = getPosition.dst2(new Vector2(nearest.get.x, nearest.get.y))
    var list = computeNeighbours.get
    import scala.collection.JavaConversions._
    for (v <- graph.vertexSet()/*list*/) {
      val distance = v.dst2(getPosition)
      if (distance < minDistance) {
        nearest = Some[Vector2](v)
        minDistance = distance
      }
    }

    (currentDestination, nearest) match {
      case (Some(cd), Some(n)) if cd.equals(n) =>
        discoverNewVertex(nearest)
        previousNode = currentNode
        currentNode = nearest
        currentDestination = selectPriorityDestination
        setNewDestination(currentDestination.get)
      case _ =>
    }
    if (!computeNeighbours.forall(p => p.contains(getCurrentDestination))) {
      currentDestination = selectRandomDestination()
      setNewDestination(currentDestination.get)
    }
    nearest
  }

  def getCurrentDestination: Vector2 = currentDestination.getOrElse(new Vector2())

  private def log = "Agent " + id + ": "

  override def equals(o: scala.Any): Boolean = {
    o match {
      case other: BaseCharacter => other.getId == id
      case _ => false
    }
  }
}

case class Guard(vector2: Vector2, id : Int) extends BaseCharacter(vector2, id){

  private[this] var fugitive: Option[Fugitive] = None

  override def getTarget: Option[Target[BaseCharacter]] = fugitive

  def hasTarget: Boolean = fugitive.isDefined

  def setFugitiveTarget(f: Option[Fugitive]): Unit = fugitive = f

}

case class Thief(vector2: Vector2, id : Int) extends BaseCharacter(vector2, id){

  private var pursuer: Option[Pursuer] = None

  private var hasReachedExit_ = false
  private var gotCaughtByGuard_ = false

  def gotCaughtByGuard: Boolean = gotCaughtByGuard_

  def hasReachedExit: Boolean = hasReachedExit_

  def setReachedExit(f: Boolean): Unit = hasReachedExit_ = f

  def setGotCaughtByGuard(f: Boolean): Unit = gotCaughtByGuard_ = f

  def hasTarget: Boolean = pursuer.isDefined

  def setPursuerTarget(p: Option[Pursuer]): Unit = pursuer = p

  override def getTarget: Option[Target[BaseCharacter]] = pursuer
}

